src/storage/mvcc/reader.rs

```diff
@@ -0,0 +1,187 @@
+            cursor.get(&k).map(|x| x.map(|x| x.to_vec())).map_err(Error::from)
+        } else {
+            self.snapshot.get(&k).map_err(Error::from)
+        }
+    }
+
+    pub fn load_lock(&mut self, key: &Key) -> Result<Option<Lock>> {
+        if let Some(ref mut cursor) = self.lock_cursor {
+            match try!(cursor.get(&key)) {
+                Some(v) => Ok(Some(try!(Lock::parse(v)))),
+                None => Ok(None),
+            }
+        } else {
+            match try!(self.snapshot.get_cf(CF_LOCK, &key)) {
+                Some(v) => Ok(Some(try!(Lock::parse(&v)))),
+                None => Ok(None),
+            }
+        }
+    }
+
+    pub fn seek_write(&mut self, key: &Key, ts: u64) -> Result<Option<(u64, u64)>> {
+        self.seek_write_impl(key, ts, false)
+    }
+
+    pub fn reverse_seek_write(&mut self, key: &Key, ts: u64) -> Result<Option<(u64, u64)>> {
+        self.seek_write_impl(key, ts, true)
+    }
+
+    fn seek_write_impl(&mut self, key: &Key, ts: u64, reverse: bool) -> Result<Option<(u64, u64)>> {
+        let (k, v) = {
+            if self.write_cursor.is_none() {
+                self.write_cursor = Some(try!(self.snapshot.iter_cf(CF_WRITE)));
+            }
+            let mut cursor = self.write_cursor.as_mut().unwrap();
+            let ok = if reverse {
+                try!(cursor.reverse_seek(&key.append_ts(ts)))
+            } else {
+                try!(cursor.seek(&key.append_ts(ts)))
+            };
+            if !ok {
+                return Ok(None);
+            }
+            (cursor.key().to_vec(), cursor.value().to_vec())
+        };
+        let (k, s, c) = try!(self.decode_write(k, v));
+        if &k == key {
+            Ok(Some((s, c)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn decode_write(&self, key: Vec<u8>, value: Vec<u8>) -> Result<(Key, u64, u64)> {
+        let k = Key::from_encoded(key);
+        let commit_ts = try!(k.decode_ts());
+        let key = try!(k.truncate_ts());
+        let start_ts = try!(value.as_slice().decode_var_u64());
+        Ok((key, start_ts, commit_ts))
+    }
+
+    pub fn get(&mut self, key: &Key, ts: u64) -> Result<Option<Value>> {
+        // Check for locks that signal concurrent writes.
+        if let Some(lock) = try!(self.load_lock(key)) {
+            if lock.ts <= ts {
+                // There is a pending lock. Client should wait or clean it.
+                return Err(Error::KeyIsLocked {
+                    key: try!(key.raw()),
+                    primary: lock.primary,
+                    ts: lock.ts,
+                });
+            }
+        }
+        match try!(self.seek_write(key, ts)) {
+            Some((start_ts, _)) => self.load_data(key, start_ts),
+            None => Ok(None),
+        }
+    }
+
+    pub fn get_txn_commit_ts(&mut self, key: &Key, start_ts: u64) -> Result<Option<u64>> {
+        match try!(self.reverse_seek_write(key, start_ts)) {
```

Comment:
For a committed transaction, its `commit_ts` is always greater than `start_ts`, and no other transaction's `commit_ts` could be between them. So `reverse_seek(start_ts)` on write CF will return the txn's commit entry (if it exists).